### DISCUSSION TITLE
Start up docker container without executing any skeema command

### DISCUSSION THREAD

**Discussion by VanTanev** [skeema/skeema#234] [Ideas and Feedback]
Hello,

We're looking to optimize our usage of Skeema in CI. Right now, we have a window of about 20-30 seconds before we need to run a skeema command, and we wanted to utilize that time to download and start skeema's MySQL 8.0 container, and keep it running and ready for the real skeema command.

Could you provide guidance on if and how something like that can be done?

Thank you!

**Comment by VanTanev** [comment]
We figured out that doing the following works out well for our use case:
```bash
$ skeema format --workspace=docker --docker-cleanup=none 
```

This will start the skeema container and keep it around, and in practice will never actually change formatting of skeema tables. We do end up doing some extra work that could be skipped, but again, in practice this is fine in our use-case.

**Comment by evanelias** [comment]
Great question. Yes indeed, that workaround is a good one. In theory you could speed it up further by running it in a dummy subdir that only contains a .skeema file and one .sql file containing a single simple CREATE... might be overkill though.

A few alternatives, just to document for posterity since this may be a good future blog post topic. In no particular order:

* If the Docker engine on your CI box doesn't already have the MySQL 8 Docker image, then that image download time alone is likely a good chunk of the latency. Running `docker pull mysql:8.0` upfront could be an easy win in that case. Only downside is you'll need to manually keep this in sync with your Skeema flavor option, if you ever upgrade to MySQL 8.1+.

* You could pre-create a mysql 8 container called "skeema-mysql-8.0", no root password, listening on 127.0.0.1 localhost loopback interface. Skeema will use that as-is if it exists (see middle of the [workspace doc](https://www.skeema.io/docs/options/#workspace))

* If your CI is only interacting with dev/test environment DBs and not real prod DBs, then workspace=docker may be unnecessary in your CI environment, and the default of workspace=temp-schema may be better. That can also optionally be combined with the [temp-schema-environment](https://www.skeema.io/docs/options/#temp-schema-environment) premium option to place the workspace on any arbitrary pre-existing database.

* The next Skeema release (v1.11.1, should be out within the next couple weeks) will include several perf improvements, including two that are specific to workspace=docker:
  * Newly-created containers will now use a bit less memory and disk space, and will take slightly less time to create (22d4b9b)
  * On Linux, when using docker-cleanup=destroy, newly-created containers will use a tmpfs data directory mount to avoid disk writes entirely (766efc3)

**Comment by VanTanev** [comment]
In our testing, I think that docker performed better than using a real server (remote aws rds instance in our case). I'll check again. 

Additionally, some time back we had issues if multiple concurrent executions of skeema were trying to use the same temp-skeema-environment. Has that limitation been lifted over the past year or so?

My guess was that to work around concurrency, I would need to generate unique temp schema names on the fly (eg, by appending timestamp).

**Comment by evanelias** [comment]
Ah, I see. Yes, if the real DB is over a WAN connection, then workspace=docker will almost always be faster.

Concurrent execution against the same workspace instance DB does indeed still require overriding the --temp-schema name with a random or semi-random value. Otherwise they'll queue up behind the lock and execute serially, and some may time out with an error, depending on the level of concurrency and the number of CREATE statements involved. So no recent functionality changes there, although the error-logging for that situation was improved in Skeema v1.10.1.

**Comment by evanelias** [comment]
Just a quick follow-up to say Skeema v1.11.1 is now available, including several perf improvements for workspace=docker. See middle of [release notes](https://github.com/skeema/skeema/releases/tag/v1.11.1) for details.

**Comment by VanTanev** [comment]
Love to see it!
